drivers/dma/xdmac_sam.c

```diff
@@ -0,0 +1,357 @@
+			| XDMAC_CC_DAM_INCREMENTED_AM;
+		break;
+	case MEMORY_TO_PERIPHERAL:
+		channel_cfg.cfg =
+			  XDMAC_CC_TYPE_PER_TRAN
+			| XDMAC_CC_CSIZE(burst_size)
+			| XDMAC_CC_DSYNC_MEM2PER
+			| XDMAC_CC_SAM_INCREMENTED_AM
+			| XDMAC_CC_DAM_FIXED_AM;
+		break;
+	case PERIPHERAL_TO_MEMORY:
+		channel_cfg.cfg =
+			  XDMAC_CC_TYPE_PER_TRAN
+			| XDMAC_CC_CSIZE(burst_size)
+			| XDMAC_CC_DSYNC_PER2MEM
+			| XDMAC_CC_SAM_FIXED_AM
+			| XDMAC_CC_DAM_INCREMENTED_AM;
+		break;
+	default:
+		SYS_LOG_ERR("'channel_direction' value %d is not supported",
+			    cfg->channel_direction);
+		return -EINVAL;
+	}
+
+	channel_cfg.cfg |=
+		  XDMAC_CC_DWIDTH(data_size)
+		| XDMAC_CC_SIF_AHB_IF1
+		| XDMAC_CC_DIF_AHB_IF1
+		| XDMAC_CC_PERID(cfg->dma_slot);
+	channel_cfg.ds_msp = 0;
+	channel_cfg.sus = 0;
+	channel_cfg.dus = 0;
+	channel_cfg.cie =
+		  (cfg->complete_callback_en ? XDMAC_CIE_BIE : XDMAC_CIE_LIE)
+		| (cfg->error_callback_en ? XDMAC_INT_ERR : 0);
+
+	ret = xdmac_sam_channel_configure(dev, channel, &channel_cfg);
+	if (ret < 0) {
+		return ret;
+	}
+
+	dev_data->dma_channels[channel].callback = cfg->dma_callback;
+
+	memset(&transfer_cfg, 0, sizeof(transfer_cfg));
+	transfer_cfg.sa = cfg->head_block->source_address;
+	transfer_cfg.da = cfg->head_block->dest_address;
+	transfer_cfg.ublen = cfg->head_block->block_size >> data_size;
+
+	ret = xdmac_sam_transfer_configure(dev, channel, &transfer_cfg);
+
+	return ret;
+}
+
+int xdmac_sam_transfer_start(struct device *dev, u32_t channel)
+{
+	Xdmac *const xdmac = DEV_CFG(dev)->regs;
+
+	if (channel >= DMA_CHANNELS_NO)
+		return -EINVAL;
+
+	/* Check if the channel is enabled */
+	if (xdmac->XDMAC_GS & (XDMAC_GS_ST0 << channel))
+		return -EBUSY;
+
+	/* Enable channel interrupt */
+	xdmac->XDMAC_GIE = XDMAC_GIE_IE0 << channel;
+	/* Enable channel */
+	xdmac->XDMAC_GE = XDMAC_GE_EN0 << channel;
+
+	return 0;
+}
+
+int xdmac_sam_transfer_stop(struct device *dev, u32_t channel)
+{
+	Xdmac *const xdmac = DEV_CFG(dev)->regs;
+
+	if (channel >= DMA_CHANNELS_NO)
+		return -EINVAL;
+
+	/* Check if the channel is enabled */
+	if (!(xdmac->XDMAC_GS & (XDMAC_GS_ST0 << channel)))
+		return 0;
+
+	/* Disable channel */
+	xdmac->XDMAC_GD = XDMAC_GD_DI0 << channel;
+	/* Disable channel interrupt */
+	xdmac->XDMAC_GID = XDMAC_GID_ID0 << channel;
+	/* Disable all channel interrupts */
+	xdmac->XDMAC_CHID[channel].XDMAC_CID = 0xFF;
+	/* Clear the pending Interrupt Status bit(s) */
+	(void)xdmac->XDMAC_CHID[channel].XDMAC_CIS;
+
+	return 0;
+}
+
+static int xdmac_sam_initialize(struct device *dev)
+{
+	const struct xdmac_sam_dev_cfg *const dev_cfg = DEV_CFG(dev);
+	Xdmac *const xdmac = dev_cfg->regs;
+
+	/* Configure interrupts */
+	dev_cfg->irq_config();
+
+	/* Enable module's clock */
+	soc_pmc_peripheral_enable(dev_cfg->periph_id);
+
+	/* Disable all channels */
+	xdmac->XDMAC_GD = UINT32_MAX;
+	/* Disable all channel interrupts */
+	xdmac->XDMAC_GID = UINT32_MAX;
+
+	/* Enable module's IRQ */
+	irq_enable(dev_cfg->irq_id);
```

Comment:
irq_enable() is usually called in relevant irq_config, with a Kconfig (or soc.h) based IRQ number (which saves a bit of ROM in device's data structure). Unless you've got a good reason to do otherwise.

Reply:
I placed `irq_enable()` inside initialization function to ensure that `IRQ_CONNECT` is the only item called by dma0_sam_irq_config. I heard several times already that we should re-implement IRQ_CONNECT macro in a way that it does not have to be called within a function. In this case we could get rid of dma0_sam_irq_config function and `dev_cfg->irq_config()` call.

If this is ever done updating the driver will be much easier if we keep the current arrangement. But maybe I'm going to far? Most other drivers do call `irq_enable()` within IRQ configuration function.

Reply:
>  I heard several times already that we should re-implement IRQ_CONNECT macro in a way that it does not have to be called within a function

Heard from who, exactly?
IRQ_CONNECT() has a runtime component. You can't pull it out of a function body. It does more that just declare data.

Reply:
Last time it was @andyross who stated in PR #1426
> Honestly IRQ_CONNECT() needs some rethinking as a framework.

It's not such a big issue but if we did not have to define a per driver instance IRQ configuration function our life would be slightly easier.

Reply:
> irq_enable() is usually called in relevant irq_config, with a Kconfig (or soc.h) based IRQ number (which saves a bit of ROM in device's data structure)

I run rom_report. At present dma_sam_xdmac.c takes 670 bytes of ROM. After placing irq_enable() in dma0_sam_irq_config() it took 676 bytes of ROM. So the current code actually generates a smaller image, probably because XDMAC_IRQn is stored here as u8_t next to periph_id which is also u8_t. Otherwise this two values would each occupy 4 bytes.

I'll leave this code as it is.